FactoryTest/FactoryTest.ino

```diff
@@ -0,0 +1,342 @@
+    }
+}
+
+void exerciseVibrator() {
+    for (int i = 0; i < 3; i += 1) {
+        digitalWrite(VIBRATOR_PIN, HIGH);
+        delay(DELAY_500ms);
+        digitalWrite(VIBRATOR_PIN, LOW);
+        delay(DELAY_500ms);
+    }
+}
+
+void exerciseBatteryVoltage() {
+    // The battery readings we expect for minimum and maximum battery voltages.
+    // TODO: we probably need to get rid of these hard-coded constants, and use a better way to derive these numbers.
+    const int readingAt12Volts = 386;
+    const int readingAt24Volts = 784;
+
+    // For the next 10 seconds we will display the battery voltage on the LEDs.
+    // Empty battery = 1 LEDs. Full battery = 7 LEDs.
+    // As you change the input voltage, the LEDs will update accordingly.
+    unsigned long startTime = millis();
+    while (millis() - startTime < 10000) {
+        // Read the current battery voltage and limit the value to the expected range.
+        uint16_t reading = analogRead(BATTERY_VOLTAGE_PIN);
+        if (reading < readingAt12Volts) reading = readingAt12Volts;
+        if (reading > readingAt24Volts) reading = readingAt24Volts;
+
+        // Calculate how full the battery is. This will be a number between 0 and 1.
+        // (Note: we use floating point because it's easier than trying to do this in fixed point. The program memory impact appears to be negligible.)
+        float fullness = float(reading - readingAt12Volts) / float(readingAt24Volts - readingAt12Volts);
+
+        // Calculate how many of the 7 LEDs we should show. 
+        uint16_t howManyLEDs = int(fullness * 6.0) + 1;
+
+        // Display the calculated number of LEDs.
+        allLEDsOff();
+        for (int i = 0; i < howManyLEDs; i += 1) {
+            digitalWrite(LEDpins[i], LOW);
+        }
+    }
+}
+
+/********************************************************************
+ * Main program that drives the sequence of exercises.
+ ********************************************************************/
+
+Exercise currentExercise;
+
+// Find the next exercise value. if e is the last exercise, then we wrap around to the first exercise.
+Exercise nextExercise(Exercise e) {
+    e = static_cast<Exercise>(static_cast<int>(e) + 1);
+    if (e == Exercise::end) {
+        e = begin;
+    }
+    return e;
+}
+
+// Do the specified exercise.
+void doExercise(Exercise exercise) {
+    currentExercise = exercise;
+
+    startExercise();
+    switch (exercise) {
+    case begin:
+        exerciseAllLEDs();
+        break;
+    case buzzer:
+        exerciseBuzzer();
+        break;
+    case vibrator:
+        exerciseVibrator();
+        break;
+    case fanMaximum:
+        exerciseFan(FAN_DUTYCYCLE_MAXIMUM);
+        break;
+    case fanMedium:
+        exerciseFan(FAN_DUTYCYCLE_MEDIUM);
+        break;
+    case fanMinimum:
+        exerciseFan(FAN_DUTYCYCLE_MINIMUM);
+        break;
+    case batteryVoltage:
+        exerciseBatteryVoltage();
+        break;
+    case fanRPM:
```

Comment:
Any issues if we hit an invalid exercise?

Reply:
I simplified the code - it is no longer possible to get an invalid value.